set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java

```diff
@@ -227,6 +226,92 @@ int validatePosition(String position, List<String> items, int pinnedItems) {
     }
     return positionInt;
   }
+  
+  int calculatePosition(int position, List<String> items) {
+    int positionFinal = items.size();
+    if (position >= 0 && position < items.size()) {
+      positionFinal = position;
+    }
+    return positionFinal;
+  }
+
+  public UserSet insertMultipleItems(List<String> items, String position, int itemsPosition, UserSet existingUserSet) 
+      throws ItemValidationException {
+    
+    validateItemsStrings(items);
+    
+    // check if the position is "pin" and is a EntityBestItem set then
+    // insert the item at the 0 position
+    UserSet userSet;
+
+    if (WebUserSetModelFields.PINNED_POSITION.equals(position)) {
+      userSet=updateItemsFromPinned(existingUserSet, items);
+    } else {
+      userSet=updateItemsFromUnpinned(existingUserSet, items, itemsPosition);
+    }
+    
+    return userSet;
+  }  
+  
+  private UserSet updateItemsFromPinned(UserSet existingUserSet, List<String> items) {
+    List<String> usersetItems=existingUserSet.getItems();
+    if(usersetItems==null) {
+      usersetItems=new ArrayList<>();
+    }
+    else {    
+      /*remove all duplicate items from the set and count the number 
+      of removed pinned items, to change the pinned field
+      */
+      for(String newItem : items) {
```

Comment:
not implemented according to the specs, we need to implement a method which computes the list of dupplicated items first and than to update the list

Reply:
It is implemented according to the specs, the method is finding the duplicates using a foor loop, not the streams, because we also need to update the pinned field if the duplicated items were pinned before, and that is all done in one foor loop.